### ISSUE TITLE
TF 2.0 issue with custom yolo training

### ISSUE THREAD

**Issue by Draghtnod** [OlafenwaMoses/ImageAI#263]
Greetings,
I have ported the library to TF 2.0 and it's integrated tensorflow.keras API. 

While I was fixing removed TF functions, i stumbled accross this lline, which was causing issues:
https://github.com/OlafenwaMoses/ImageAI/blob/ca450c093225a97a681e922701cb52dc3721029f/imageai/Detection/Custom/yolo.py#L43
The comment suggests, that `batch_seen` counts the batches the Layer has seen, but it gets reset to 0 every time the Layer is called. So the following condition 
https://github.com/OlafenwaMoses/ImageAI/blob/ca450c093225a97a681e922701cb52dc3721029f/imageai/Detection/Custom/yolo.py#L151
is always true. Am I right or just stupid?  :)

I moved the definition of  `batch_seen` to `__init__` and made it `trainable=False`. Now the training seems to work again. At least the loss decreases. I'll need a few more hours to be sure.

**Comment by Draghtnod** [comment]
Looks like the Model breaks as soon as the above condition gets false.
This is the log with batch_size increasing and the original condition:
```
Generating anchor boxes for training images and annotation...
Average IOU for 9 anchors: 0.99
Anchor Boxes generated.
Detection configuration saved in  data\tmp\10044\json\detection_config.json
Training on: 	['feet', 'head']
Training with Batch Size:  1
Number of Experiments:  200
Warmup Batches: 10680
Epoch 1/203
3560/3560 - 961s - loss: 15.1387 - yolo_layer_loss: 7.5839 - yolo_layer_1_loss: 2.5836 - yolo_layer_2_loss: 4.9712 - val_loss: 9.4518 - val_yolo_layer_loss: 6.7878 - val_yolo_layer_1_loss: 0.8804 - val_yolo_layer_2_loss: 1.7835
Epoch 2/203
3560/3560 - 874s - loss: 6.8438 - yolo_layer_loss: 5.1568 - yolo_layer_1_loss: 0.5745 - yolo_layer_2_loss: 1.1125 - val_loss: 9.0644 - val_yolo_layer_loss: 7.4237 - val_yolo_layer_1_loss: 0.5768 - val_yolo_layer_2_loss: 1.0639
Epoch 3/203
3560/3560 - 878s - loss: nan - yolo_layer_loss: nan - yolo_layer_1_loss: nan - yolo_layer_2_loss: nan - val_loss: nan - val_yolo_layer_loss: nan - val_yolo_layer_1_loss: nan - val_yolo_layer_2_loss: nan
```
This is an identical run, but with the above condition hardcoded to true:
```
Generating anchor boxes for training images and annotation...
Average IOU for 9 anchors: 0.99
Anchor Boxes generated.
Detection configuration saved in  data\tmp\5392\json\detection_config.json
Training on: 	['feet', 'head']
Training with Batch Size:  1
Number of Experiments:  200
Warmup Batches: 10680
Epoch 1/203
3560/3560 - 987s - loss: 15.2869 - yolo_layer_loss: 7.7544 - yolo_layer_1_loss: 2.5952 - yolo_layer_2_loss: 4.9374 - val_loss: 14.5311 - val_yolo_layer_loss: 9.7359 - val_yolo_layer_1_loss: 2.4219 - val_yolo_layer_2_loss: 2.3732
Epoch 2/203
3560/3560 - 888s - loss: 6.7059 - yolo_layer_loss: 5.1175 - yolo_layer_1_loss: 0.5176 - yolo_layer_2_loss: 1.0707 - val_loss: 11.5325 - val_yolo_layer_loss: 7.7341 - val_yolo_layer_1_loss: 1.7149 - val_yolo_layer_2_loss: 2.0835
Epoch 3/203
3560/3560 - 856s - loss: 5.6355 - yolo_layer_loss: 4.5523 - yolo_layer_1_loss: 0.3223 - yolo_layer_2_loss: 0.7609 - val_loss: 19.9782 - val_yolo_layer_loss: 15.2007 - val_yolo_layer_1_loss: 2.3002 - val_yolo_layer_2_loss: 2.4773
Epoch 4/203
3560/3560 - 868s - loss: 5.1894 - yolo_layer_loss: 4.2727 - yolo_layer_1_loss: 0.2744 - yolo_layer_2_loss: 0.6423 - val_loss: 28.8574 - val_yolo_layer_loss: 24.2829 - val_yolo_layer_1_loss: 2.4340 - val_yolo_layer_2_loss: 2.1405
Epoch 5/203
3560/3560 - 866s - loss: 4.8808 - yolo_layer_loss: 4.0858 - yolo_layer_1_loss: 0.2431 - yolo_layer_2_loss: 0.5519 - val_loss: 26.8364 - val_yolo_layer_loss: 23.2204 - val_yolo_layer_1_loss: 2.2337 - val_yolo_layer_2_loss: 1.3823
Epoch 6/203
3560/3560 - 863s - loss: 4.6883 - yolo_layer_loss: 3.9706 - yolo_layer_1_loss: 0.2306 - yolo_layer_2_loss: 0.4871 - val_loss: 25.4950 - val_yolo_layer_loss: 21.6413 - val_yolo_layer_1_loss: 2.3024 - val_yolo_layer_2_loss: 1.5513
Epoch 7/203
3560/3560 - 890s - loss: 4.5435 - yolo_layer_loss: 3.8887 - yolo_layer_1_loss: 0.2192 - yolo_layer_2_loss: 0.4356 - val_loss: 40.2588 - val_yolo_layer_loss: 37.5845 - val_yolo_layer_1_loss: 1.5578 - val_yolo_layer_2_loss: 1.1165
```

**Comment by Draghtnod** [comment]
I figured it out. Variables behave a bit different in TF 2.0. Custom Training and Inferencing with TF 2.0 works now.

The actual issue of the inf-loss was a zero division within the YOLO Layers, caused by 0 anchors. The 0 anchors were caused by all annotation boxes having an identical size. Is that a bug or a feature?

@OlafenwaMoses If you are interested in my branch, let me know. It might help you porting the other parts to TF 2.0.
